### PR TITLE
Activation offloading for fullfinetuning + fix tied embedding

### PR DESCRIPTION
**Pull request by felipemello1**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

#### Changelog
- Added activation offloading for single/distributed full finetuning
- Found that we didnt address the tied embedding issue. Fixed it by adding a dummy nn.Module so that tensor hooks would work with it
- Moved the context manager logic to a function
- Made the recipes more aligned

#### Test plan
image
image
image
image

### CODE REVIEW

File: recipes/full_finetune_single_device.py [hunk 1]
```diff
@@ -517,13 +557,15 @@ def _setup_data(
             sampler=sampler,
             # dropping last avoids shape issues with compile + flex attention
             drop_last=True,
-            collate_fn=partial(
```
Comment (felipemello1) [inline recipes/full_finetune_single_device.py:520]: didnt touch this. It was pre commit hook

File: recipes/lora_finetune_distributed.py [hunk 2]
```diff
@@ -151,16 +146,6 @@ def __init__(self, cfg: DictConfig) -> None:
...
 
-        # training attributes
-        self._enable_activation_checkpointing = cfg.enable_activation_checkpointing
-        self._enable_activation_offloading = cfg.get(
-            "enable_activation_offloading", False
-        )
-        if self._enable_activation_offloading and self._device.type != "cuda":
-            raise RuntimeError(
-                "enable_activation_offloading should only be enabled for training on CUDA"
-            )
```
Comment (felipemello1) [inline recipes/lora_finetune_distributed.py:162]: removed from init. This is handled in setup_model

File: recipes/full_finetune_distributed.py [hunk 3]
```diff
@@ -222,9 +239,11 @@ def setup(self, cfg: DictConfig) -> None:
...
         self._optimizer = self._setup_optimizer(
             cfg_optimizer=cfg.optimizer,
-            opt_state_dict=checkpoint_dict[training.OPT_KEY]
-            if self._resume_from_checkpoint
-            else None,
+            opt_state_dict=(
+                checkpoint_dict[training.OPT_KEY]
+                if self._resume_from_checkpoint
+                else None
+            ),
```
Comment (felipemello1) [inline recipes/full_finetune_distributed.py:246]: pre commit hook

File: recipes/full_finetune_distributed.py [hunk 4]
```diff
@@ -509,13 +549,15 @@ def _setup_data(
...
-            else padded_collate_packed,
+            collate_fn=(
+                partial(
+                    collate_fn,
+                    padding_idx=self._tokenizer.pad_id,
+                    ignore_idx=self._loss_fn.ignore_index,
+                )
+                if not packed
+                else padded_collate_packed
+            ),
```
Comment (felipemello1) [inline recipes/full_finetune_distributed.py:560]: pre commit hook

File: recipes/full_finetune_distributed.py [hunk 5]
```diff
@@ -440,6 +460,26 @@ def _is_layer_fqn(s: str) -> bool:
...
+                )
+            if not enable_activation_checkpointing:
+                raise RuntimeError(
+                    "enable_activation_offloading should only be True when enable_activation_checkpointing is True"
+                )
+        elif enable_activation_checkpointing:
+            log.info(
+                "Hint: enable_activation_checkpointing is True, but enable_activation_offloading isn't. "
+                "Enabling activation offloading should reduce memory further."
+            )
```
Comment (felipemello1) [inline recipes/full_finetune_distributed.py:477]: the only think that i dont like about this is that it could fail much faster if we added it to the __init__. But i like that the checks are near the code where it matters. So i am not sure which one to pick.

File: recipes/full_finetune_distributed.py [hunk 6]
```diff
@@ -440,6 +460,26 @@ def _is_layer_fqn(s: str) -> bool:
...
+
+        self.activations_handling_ctx = training.get_act_offloading_ctx_manager(
+            model, enable_activation_offloading
+        )
```
Comment (felipemello1) [inline recipes/full_finetune_distributed.py:481]: added function to handle NoOp / getting context

File: recipes/full_finetune_distributed.py [hunk 7]
```diff
@@ -440,6 +460,26 @@ def _is_layer_fqn(s: str) -> bool:
...
 
+        # activation checkpointing/offloading
+        if enable_activation_offloading:
+            if self._device.type != "cuda":
+                raise RuntimeError(
+                    "enable_activation_offloading should only be True for training on CUDA"
+                )
+            if not enable_activation_checkpointing:
+                raise RuntimeError(
+                    "enable_activation_offloading should only be True when enable_activation_checkpointing is True"
```
Comment (felipemello1) [inline recipes/full_finetune_distributed.py:471]: no point in running offloading when AC is off. Its extremely slow.

File: torchtune/modules/tied_linear.py [hunk 8]
```diff
@@ -9,13 +9,33 @@
 import torch.nn.functional as F
 
 
+class Linear(nn.Module):
```
Comment (felipemello1) [inline torchtune/modules/tied_linear.py:12]: docstring explains why i had to add this

**Comment by felipemello1** [comment]
Will add the option to all the configs for full finetuning

File: recipes/full_finetune_distributed.py (hunk 5 above)
Comment (janeyx99) [reply recipes/full_finetune_distributed.py:477]: I think failing much faster is better

File: torchtune/modules/tied_linear.py (hunk 8 above)
Comment (ebsmothers) [reply torchtune/modules/tied_linear.py:12]: Sorry I'm a bit confused by this change on two fronts: 

(1) Does this not change the key names in the state dict?
(2) Now that we have a module again, how do we not wind up right back where we started with to_empty?

File: recipes/configs/llama3_2_vision/11B_full_single_device.yaml [hunk 9]
```diff
@@ -68,6 +68,7 @@ device: cuda
 
 # Memory management
 enable_activation_checkpointing: True
+enable_activation_offloading: False  # True reduces memory
```
[the remark this answers is not in the document]
Comment (felipemello1) [reply recipes/configs/llama3_2_vision/11B_full_single_device.yaml:71]: i didnt include it in my tests. Let me do it tomorrow.

File: recipes/full_finetune_distributed.py [hunk 10]
```diff
@@ -44,13 +44,25 @@ class FullFinetuneRecipeDistributed(FTRecipeInterface):
...
 
+        - Activation Offloading. This can be controlled using the ``enable_activation_offloading``
+            flag. Activation offloading is a technique similar to activations checkpointing that helps
+            reduce the memory footprint to prevent OOMs on CUDA and enable bigger batches. Where activations
+            checkpointing drops the activation in the forward to recompute it later in the backward,
+            activations offloading will drop the activation in the forward to the CPU and bring it
+            back during the backward pass. As always, there is a tradeoff--these savings in memory can
+            come at the cost of training performance and CPU resources. To recover some runtime cost,
+            we've added an option to enable offloading on a different stream to permit overlapping with
+            the computation. This option is currently only available on PyTorch nightly 2.5.0.dev20240907
```
[the remark this answers is not in the document]
Comment (felipemello1) [reply recipes/full_finetune_distributed.py:62]: agreed

File: recipes/lora_finetune_distributed.py (hunk 2 above)
Comment (felipemello1) [reply recipes/lora_finetune_distributed.py:162]: I added it back after this:

image

File: torchtune/modules/tied_linear.py (hunk 8 above)
Comment (felipemello1) [reply torchtune/modules/tied_linear.py:12]: (1) No. The Linear nn.Module doesnt have any weights. The weight is passed in the forward only. 
```
Class Linear(nn.Module):
      def forward(x, weight):
             return F.Linear(x, weight)
```

TiedLinear is still a regular python class, and the key name is still model.output.tied_module.weight --> translates to --> model.TiedLinear.tok_embedding.weight

(2) i tested it with FSDP, and confirmed that the memory pointers are the same in model.tok_embedding.weight and model.output.weight. So things are fine. Is that what you were referencing too?

File: tests/recipes/test_full_finetune_distributed.py [hunk 11]
```diff
@@ -34,6 +34,7 @@ def _get_test_config_overrides(self):
             "batch_size=4",
             "dtype=fp32",
             "enable_activation_checkpointing=False",
+            "enable_activation_offloading=False",
```
[the remark this answers is not in the document]
Comment (felipemello1) [reply tests/recipes/test_full_finetune_distributed.py:37]: we should. I will see if i can quickly add it tomorrow

File: torchtune/training/_activation_offloading.py [hunk 12]
```diff
@@ -327,3 +331,73 @@ def noop(tensor):
...
+            log.warning(
+                "During activation offloading, no output head was detected. "
+                "If your model has an output head, it will be offloaded. "
+                "This usually greatly slows training, given the large vocabulary size. "
+                "To change this behavior, set your output head as model.output and make it "
+                "an nn.Module."
+            )
+
+    else:
+        activations_handling_ctx = contextlib.nullcontext()
```
Comment (felipemello1) [inline torchtune/training/_activation_offloading.py:401]: this covers all of our cases, but it looks horrible. I think its fine to ship like it, but in the future change it to identify tensor size, and if larger than threshold, make it a non op

File: tests/recipes/test_lora_finetune_distributed.py [hunk 13]
```diff
@@ -173,6 +175,7 @@ def test_training_state_on_resume(
             resume_from_checkpoint=True \
             metric_logger.filename={log_file} \
             enable_activation_checkpointing=True \
+            enable_activation_offloading=True \
```
Comment (felipemello1) [inline tests/recipes/test_lora_finetune_distributed.py:178]: I enabled it for some tests in this file only. It tests lora and distributed. Ideally, we should have it vs many other parameters, like compile, vision and tiedembeddings. I wont address those in this PR. This should be part of the testing improvement, IMO

File: torchtune/modules/tied_linear.py [hunk 14]
```diff
@@ -26,12 +46,13 @@ class TiedLinear:
...
     def __init__(self, tied_module: nn.Module):
         self.tied_module = tied_module
+        self.linear = Linear()
         if not hasattr(tied_module, "weight"):
             raise AttributeError(
                 "Provided module does not have attribute 'weight'. Please check your tied_module."
             )
 
-    def __call__(self, x: torch.Tensor) -> torch.Tensor:
+    def __call__(self, x: torch.Tensor):
```
Comment (joecummings) [inline torchtune/modules/tied_linear.py:55]: Why did you remove the type hint?

File: recipes/full_finetune_single_device.py [hunk 15]
```diff
@@ -569,7 +613,8 @@ def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
         # Shape [b, s], needed for the loss not the model
         labels = batch.pop("labels")
 
-        logits = self._model(**batch)
+        with self.activations_handling_ctx:
```
Comment (SalmanMohammadi) [inline recipes/full_finetune_single_device.py:616]: Curiosity brings me here. If I gather correctly, every model forward pass for which which we want to offload activations for needs to sit inside this context manager?

Comment (felipemello1) [reply recipes/full_finetune_single_device.py:616]: correct!

File: torchtune/training/_activation_offloading.py [hunk 16]
```diff
@@ -345,3 +352,82 @@ def noop(tensor):
...
+            )
+            # if isinstance(model.decoder, nn.Module):
+            #     model.decoder.output.register_forward_pre_hook(
+            #         lambda *args: noop_ctx.__enter__()
+            #     )
+            #     model.decoder.output.register_forward_hook(
+            #         lambda *args: noop_ctx.__exit__(), always_call=True
+            #     )
+            #     output_head_detected = True
+
```
Comment (felipemello1) [inline torchtune/training/_activation_offloading.py:420]: this needs debugging in a follow up pr

Comment (joecummings) [reply torchtune/training/_activation_offloading.py:420]: haha can you remove the commented out code?

**Review by joecummings — APPROVED** [review]
We're not savages, remove the commented out code.
